docs/proposals/20210222-machineadm.md

```diff
@@ -0,0 +1,835 @@
+</td>
+
+<tr>
+<td>machineadm-timesync-plugin-ntpd</td>
+<td>
+A plugin that configures ntpd.
+</td>
+
+<tr>
+<td>machineadm-timesync-plugin-chronyd</td>
+<td>
+A plugin that configures chronyd.
+</td>
+
+</tbody>
+</table>
+
+### Phased development
+
+With reference to the modalities described in the requirements specification, we propose 3 phases of development:
+
+| Phase             | What gets implemented           |
+| ----------------- | ------------------------------- |
+| Phase 1 - Alpha 1 | Provisioning Modality           |
+| Phase 2 - Alpha 2 | Preparation and Post Modalities |
+| Phase 3 & GA      | API Stabilisation               |
+
+### Plugin architecture
+
+Machineadm will use the [go-plugin][go-plugin] architecture used by Hashicorp. Machineadm will expect plugins to
+use consistent filename formats depending on what type of plugin it is (and will be documented in the type).
+
+Plugins can either be "on path", in a directory supported in the PATH variable, or:
+* Linux: stored in `/usr/libexec/machineadm` or `/usr/local/libexec/machineadm`
+* Windows: stored in `"C:\Program Files\machineadm\plugins"`
+
+### Security Model
+
+#### Userdata storage
+
+Machineadm's security model is intended such that secrets associated with machine configuration are stored on the
+management Kubernetes cluster and are secured to the level that the backing API server and etcd are secured.
+
+For machine bootstrapping, secrets are intended to be delivered by an infrastructure provider to a suitable secure
+location, and to also provide a mechanism for the machine to report back status. Examples include:
+
+* Amazon S3 or S3-compatible API such as Minio
+* AWS Secrets Manager and AWS Systems Manager Parameter Store
+
+When using external storage, we assume security is configured appropriately (e.g. Amazon Identity and Access Management),
+and is outside of the system boundary of machineadm bootstrap controller.
+
+In addition, some providers may require a secondary mechanism to encrypt data. For example, a deployment of an S3 compatible
+API may not support the same levels of fine-grained IAM access supported by EC2. A provider may then want to further encrypt
+the data. In this case  we will additionally support a data block for in-band binary data (see [#encrypted-configuration](#encrypted-configuration)) that can be decrypted with a key derivation algorithm and a passphrase from an external source. We do not assume a KMS can
+directly encrypt the entire machine configuration as most systems have byte limits of around 64KB. Instead we assume
+the use of envelope encryption where a KMS encrypts a random string that is itself used as a key for encryption/data
+of data.
+
+We furthermore protect the KMS issued random string by deriving a key from it using SHA-512 PBKDF2-HMAC hashed to 50,000 rounds, with the result used for AES-256-GCM. This provides FIPS compliance with a suitable level of encryption given the
```

Comment:
Rolling your own encryption is rarely advised, sops PGP will meet the needs of users without a KMS

Reply:
Using AES-GCM and PBKDF2 is "not rolling your own encryption".

Reply:
Furthermore, from experience, libgcrypt is extremely painful to use programatically, even more so from cgo. We'd have to use https://github.com/keybase/go-crypto, but it uses independent algos. Using the go crypto libraries with shims for BoringSSL more easily certifies FIPS compliance.

Reply:
sops already does all the heavy lifting for PGP, 

I don't mean rolling your own crypto algorithm, but rather how the algorithm is used - A lot can go wrong even if you are using proven algorithms - https://loup-vaillant.fr/articles/rolling-your-own-crypto / https://security.stackexchange.com/questions/18197/why-shouldnt-we-roll-our-own